Dockerfile

```diff
@@ -1,18 +1,60 @@
+ARG CONTAINER_RUNTIME=bionic
 FROM mcr.microsoft.com/dotnet/core/sdk:3.1-bionic AS build
-WORKDIR /es
+ARG RUNTIME=linux-x64
 
-COPY ./ .
-RUN dotnet restore ./src/EventStore.sln
+WORKDIR /build/ci
 
-WORKDIR /es/src
-RUN dotnet publish -c Debug -o ../../out
+COPY ./ci ./
 
-WORKDIR /out
-RUN ls -la
+WORKDIR /build/src
 
-FROM mcr.microsoft.com/dotnet/core/aspnet:3.1-bionic AS dest
-WORKDIR /eventstore
-COPY --from=build /out/ ./
+COPY ./src/EventStore.sln ./src/*/*.csproj ./src/Directory.Build.* ./
+
+RUN for file in $(ls *.csproj); do mkdir -p ./${file%.*}/ && mv $file ./${file%.*}/; done
+
+RUN dotnet restore --runtime=${RUNTIME}
+
+COPY ./src .
+
+RUN dotnet build --configuration=Release --runtime=${RUNTIME} --no-restore --framework=netcoreapp3.1
+
+FROM build as test
+ARG RUNTIME=linux-x64
+RUN echo '#!/usr/bin/env sh\n\
+find /build/src -maxdepth 1 -type d -name "*.Tests" -print0 | xargs -0 -n1 dotnet test --no-restore --blame --configuration=Release --results-directory=../testresults --verbosity=normal --logger=trx --runtime=${RUNTIME} --settings ../ci/ci.runsettings' \
+    >> /build/test.sh && \
+    chmod +x /build/test.sh
+CMD ["/build/test.sh"]
+
+FROM build as publish
+ARG RUNTIME=linux-x64
+
+RUN dotnet publish --configuration=Release --no-build --runtime=${RUNTIME} --self-contained \
+     --framework=netcoreapp3.1 --output /publish /p:PublishTrimmed=true EventStore.ClusterNode
+
+FROM mcr.microsoft.com/dotnet/core/runtime-deps:3.1-${CONTAINER_RUNTIME} AS runtime
```

Comment:
We are publishing a self contained application, we shouldn't need the 3.1 runtime.